mindmeld/components/dialogue.py

```diff
@@ -689,6 +692,189 @@ def _get_dialogue_handler(self, dialogue_state):
+
+        if slot.hints:
+            # hints / user-list validation
+            extracted_feature = {'hint_validated_entity': text} if text in slot.hints else {}
+
+        # return True iff user input results in extracted features (i.e. validated)
+        return extracted_feature != {}
+
+    def _initial_fill(self, request):
+        """
+        Performs the first pass and fills the entity form with entity values available
+        in the initial query.
+
+        Args:
+            request (Request): The request object.
+        """
+        for entity in request.entities:
+            entity_type = entity['type']
+            role = entity['role']
+            value = entity['text']
+
+            for slot in self._local_form:
+                if entity_type == slot.entity:
+                    if slot.role == None or (role == slot.role):
+                        slot.value = value
+
+    def __call__(self, request, responder):
+        """The iterative call to fill missing slots in the entity form till all slots have been
+        filled up or the flow has been exited.
+
+        Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+        """
+        if self._slot_not_prompted is None or self._local_form is None:
+            # Entering the flow
+            if not self._entity_form:
+                self._exit_flow(responder)
+                return
+
+            self._slot_not_prompted = True
+            # Fill the form with the entities in the first query
+            self._local_form = copy.deepcopy(self._entity_form)
+            self._initial_fill(request)
+
+        for slot in self._local_form:
+            entity_ = slot.entity
+            value = slot.value
+            nlr = (slot.responses if slot.responses
+                else ["Please provide value for: {}".format(entity_)]
+                )
+
+            if not value:
+                if self._slot_not_prompted:
+                    self._set_target_state(responder)
+                    responder.reply(nlr)
+                    self._slot_not_prompted = False
+                    return
+                else:
+                    if self._validate(request.text, entity_, slot):
+                        slot.value = request.text
+                        self._slot_not_prompted = True
+                    else:
+                        # retry logic
+                        if 'retry_count' in responder.frame:
+                            if responder.frame['retry_count'] < self._retry_attempts:
+                                responder.frame['retry_count'] += 1
+                            else:
+                                # max attempts exceeded, reset counter, exit flow.
+                                del responder.frame['retry_count']
+                                responder.reply(['Invalid response. How can I help you?'])
```

Comment:
We should have a way to customize this response as well

Reply:
Removed this response altogether, now it reprocesses the current query with no target dialogue state once it's reached the retry attempt limit - smoother transition to new flow/handler.